### ISSUE TITLE
Is it suspicious that I got higher performance for OCSORT against DeepOCSORT on MOT17?

### ISSUE THREAD

**Issue by Jumabek** [mikel-brostrom/yolo_tracking#1015]
### Search before asking

- [X] I have searched the Yolo Tracking [issues](https://github.com/mikel-brostrom/yolo_tracking/issues) and found no similar bug report.

I observed a benchmark [here](https://github.com/mikel-brostrom/yolo_tracking/wiki/StrongSORT-vs-BoTSORT-vs-OCSORT), but it does not incorporate Deep OCSORT. I assume this omission is due to the benchmark's age.

**Some context regarding my goal** 
I conducted a comparative experiment [here](https://github.com/Jumabek/SurvTrack) using the VIRAT dataset and found OCSORT to be the most compelling based on the metrics IDF1, MOTA, and MOTP.

I am now moving onto the second phase, which involves designing novel appearance features. The following are my initial findings.

**Question**
When I compared Deep OCSORT to OCSORT, the results were surprising. I expected Deep OCSORT, which includes an additional appearance module, to exhibit a superior HOTA score.

Specifically, results below show, OCSORT alone (w/o) apperance features is better than Deep OCSORT. 
I am wondering why and if anyone has made similar observations.

OCSORT HOTA COMBINED                           40.648
DEEP OCSORT HOTA COMBINED                 38.346

**Commands to reproduce**
``` python3 examples/val.py       --tracking-method ocsort     --benchmark MOT17 ```
``` python3 examples/val.py       --tracking-method deepocsort     --benchmark MOT17 ```

**Upcoming action**
Planning to check this performance on differeent resolution and detector confidence thresholds

**Comment by mikel-brostrom** [comment]
The config files have been evolved for a different set of models. I still have the evolve the hparams for the standard YoloX models used for evaluating MOT17 and reproduce paper results.

**Comment by mikel-brostrom** [comment]
I do not recommend to use the default configs as they have not been evolved for the current SOTA YoloX models

**Comment by Jumabek** [comment]
`I do not recommend to use the default configs as they have not been evolved for the current SOTA YoloX models`
thank you @mikel-brostrom  this is valuable, I should ensure OCSORT and deepOCSORT configs to be the same for fair comparison

**Comment by mikel-brostrom** [comment]
> I should ensure OCSORT and deepOCSORT configs to be the same for fair comparison

Absolutely, that is a first good start

**Comment by Jumabek** [comment]
As note for my future self:
I evolved OCSORT for MOT17 with following command
` python examples/evolve.py --tracking-method ocsort --benchmark MOT17 --n-trials 100  --objective HOTA`

Best was trial 74 with value: 41.318.

```
# Trial number:      74
# HOTA, MOTA, IDF1:  [41.318]
asso_func: giou
conf: 0.424880788280455
delta_t: 1
det_thresh: 0
inertia: 0.3348497441538068
iou_thresh: 0.2625783256595632
max_age: 50
min_hits: 1
use_byte: false

```

**Comment by mikel-brostrom** [comment]
What was the total number of trials you run @Jumabek?

**Comment by Jumabek** [comment]
@mikel-brostrom it was 100. 

also pls note that by default it is using yolov8n.pt

**Comment by Jumabek** [comment]
![image](https://github.com/mikel-brostrom/yolo_tracking/assets/4343873/28f3cd00-662f-4af6-988c-c76b488d65f0)

Also, I have some hypothesis that allowing lower confidence thresholds for the trial may help as 0.1 was much better than other conf thresholds here


https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=10173520&tag=1

![image](https://github.com/mikel-brostrom/yolo_tracking/assets/4343873/030269be-20b6-40ab-af77-b1c76e3d2e6d)

Especially for OCSORT, DeepOCSORT and STRONGSORT

**Comment by Jumabek** [comment]
BTW, in the previous experiments I set detection model to yolov8m.pt

**Comment by mikel-brostrom** [comment]
Performing experiments with different models I noticed that the optimal confidence for people (for maximum F1 score for example) given a model that is trained for multiple classes, is not always known... That is why confidence is searched for in the evolution process. 

Could be that for the n models, a lower `conf` gives better results as its overall performance is worse that the larger models, which also implies that it is less confident about the `conf`s of what it predicts.

**Comment by Jumabek** [comment]
@mikel-brostrom in the process i have discovered a bug. 

Code is not using conf threshold from yaml files, but instead using the default 0.5

When i set differing conf values for yaml, there is no change in result. also double checked with debugger.

**Comment by Jumabek** [comment]
@mikel-brostrom should make bug report?

**Comment by mikel-brostrom** [comment]
In `evolve.py` conf is overwritten for each trial:

https://github.com/mikel-brostrom/yolo_tracking/blob/fec13e7264bdd38cae7833217162ea256267dd36/examples/evolve.py#L56

In val you should use:

`python3 examples/val.py --conf 0.1 --tracking-method ocsort     --benchmark MOT17`

**Comment by Jumabek** [comment]
Some experiments with default [deepocsort.yaml](https://github.com/mikel-brostrom/yolo_tracking/blob/master/boxmot/configs/deepocsort.yaml) on MOT17 benchmark

Deep OCSORT yaml uses conf =0.5, but we used 0.45 (as it was default value in [val.py](https://github.com/mikel-brostrom/yolo_tracking/blob/master/examples/val.py#L339))


```
# Trial number:      0
# HOTA, MOTA, IDF1:  [27.884, 11.425, 21.545]
alpha_fixed_emb: 0.95
asso_func: iou
aw_off: false
aw_param: 0.5
cmc_off: false
**conf: 0.45**
delta_t: 3
det_thresh: 0
embedding_off: false
inertia: 0.2
iou_thresh: 0.3
max_age: 30
min_hits: 3
new_kf_off: false
w_association_emb: 0.75

```
**Metric**: (Combined) HOTA
**Variable**: Yolo models
**Results**
- yolov8n : 38.727 
- yolov8s : 41.495
- yolov8m: 41.412
- yolov8l  : 40.865 
- yolov8x : 40.805


--
change the conf threshold from 0.45 to 0.1

**Metric**: (Combined) HOTA
**Variable**: Yolo models
**Results**
- yolov8n :  31.213 
- yolov8s :  34.507 
- yolov8m:  37.484 
- yolov8l  :  40.044 
- yolov8x :  38.948

**Comment by mikel-brostrom** [comment]
Thanks for sharing! Super interesting that YOLO S & M are both superior to both L and X

**Comment by mikel-brostrom** [comment]
However, notice that the optimal conf may be different for the different models

**Comment by Jumabek** [comment]
`You would need to do a sweep on conf for knowing which is better`

working on that @mikel-brostrom

**Comment by mikel-brostrom** [comment]
Waiting patiently for the results :smile:

**Comment by Jumabek** [comment]
@mikel-brostrom here is the result of different confidence threshold for different models using this repo on DeepOCSORT 

Interestingly, larger models are more immune to varying confidence threshold. 

![image](https://github.com/mikel-brostrom/yolo_tracking/assets/4343873/3416497d-e81b-4c9e-8e36-872f37876b67)

**Comment by Jumabek** [comment]
It is noteworthy to tell that, when I ran the original deepOCSORT repo, I indeed observed HOTA of 66.428 without linear interpolation and 68 with linear interpolation.

to reproduce
python3 main.py --exp_name $exp --post --emb_off --cmc_off --aw_off --new_kf_off --grid_off --dataset mot17 
Is it because they use custom yolox detector that is trained only for person detection?

**Comment by mikel-brostrom** [comment]
> Is it because they use custom yolox detector that is trained only for person detection?

They use the same detector as ByteTrack, BotSORT. From the BoTSORT paper:

> "The detector is YOLOX [24] with YOLOX-X as the backbone and COCO-pretrained mode l[36] as the initialized weights. For MOT17, the training schedule is 80 epochs on the combination of MOT17, CrowdHuman, Cityperson and ETHZ. For MOT20 and HiEve, we only add CrowdHuman as additional training data."

Luckily these models are avaialble here as well :rocket: 

https://github.com/mikel-brostrom/yolo_tracking/blob/8f95b860c7107ae4c4092531aac418b5b76ed2eb/examples/detectors/yolox.py#L11-L17

**Comment by Jumabek** [comment]
@mikel-brostrom very valuable, thank you !

**Comment by Jumabek** [comment]
makes total sense to obtain better detections using domain specific detector
`80 epochs on the combination of MOT17, CrowdHuman, Cityperson and ETHZ. For MOT20 and HiEve`

**Comment by mikel-brostrom** [comment]
> makes total sense to obtain better detections using domain specific detector

Yup :smile:

**Comment by Jumabek** [comment]
Other experiments for OCSORT, yaml was geenrated in 100 trials with evolve.py

```
# Trial number:      74
# HOTA, MOTA, IDF1:  [41.318], verfied by val.py
asso_func: giou
conf: 0.424880788280455
delta_t: 1
det_thresh: 0
inertia: 0.3348497441538068
iou_thresh: 0.2625783256595632
max_age: 50
min_hits: 1
use_byte: false
```


![image](https://github.com/mikel-brostrom/yolo_tracking/assets/4343873/d79bda4e-2396-4ab1-892b-db529e58e38f)


conf_values=[0.6 0.5 0.45 0.4 0.35 0.3 0.25 0.2 0.1]
yolov8m = [ 41.137, 41.92, 42.712,  42.795 , 43.445, 43.271, 43.615 , 43.805, 41.338 ]
yolov8n = [
 38.601, 39.809,  40.93, 0.509 , 40.166 ,  40.966  , 40.525 , 39.873 , 35.871 
]
yolov8x = [
40.609 , 42.061,  42.042, 42.482,  42.787, 43.028 , 43.471 , 43.901, 41.995 
]

**Comment by mikel-brostrom** [comment]
What reid model did you use?

**Comment by mikel-brostrom** [comment]
>  python3 main.py --exp_name $exp --post --emb_off --cmc_off --aw_off --new_kf_off --grid_off --dataset mot17 

Notice that Gaussian linear interpolation (GSI) is available in this repo by `examples/val.py --gsi`

**Comment by Jumabek** [comment]
> Which reid model did you use? I recommend `osnetx1_0_dukemtcereid.pt` out of the lightweight models for maximum performance. 

@mikel-brostrom sorry I missed your query, I used default `mobilenetv2_x1_4_dukemtmcreid`


> Make no sense at all that OCSORT is better than DeepOCSORT. This is because DeepOCSORT is basically the same as OCSORT but with added appearance information in the association process, compared to OCSORT which only uses motion. This should give DeepOCSORT a slightly higher IDF1 score and hence a higher HOTA.

That is exactly the reason why I am surprised. From below we can see that DeepOCSORT has low Association Accuracy (AssA). But it was supposed to increased AssA of OCSORT. 

![image](https://github.com/mikel-brostrom/yolo_tracking/assets/4343873/82dbf750-2c8a-42a4-8499-03b96e754281)

**Comment by mikel-brostrom** [comment]
I may need to revisit DeepOCSORT. I saw better performance when setting the embedding weight to 0... When generating metrics for the README of this repo

**Comment by mikel-brostrom** [comment]
I fixed the issue based on the metrics I get.

**Comment by Jumabek** [comment]
@mikel-brostrom thanks, I confirm the improvement of DeepOCSORT on MOT17 dataset
![image](https://github.com/mikel-brostrom/yolo_tracking/assets/4343873/54570f22-530d-496e-a437-7b14d3e43518)

**Comment by mikel-brostrom** [comment]
Great!! Thanks for the feedback @Jumabek

**Comment by Jumabek** [comment]
@mikel-brostrom yes n YOLOX_m model,

 any idea why OCSORT and DeepOCSORT has signifcantly low performance compared to other trackers ?

**Comment by Jumabek** [comment]
OCSORT yaml
```
# Trial number:      137
# HOTA, MOTA, IDF1:  [55.567]
asso_func: giou
conf_thres: 0.5122620708221085
delta_t: 3
det_thresh: 0
inertia: 0.2
iou_thresh: 0.3
max_age: 30
min_hits: 1
use_byte: false

```


DeepOCSORT

```
# Trial number:      0
# HOTA, MOTA, IDF1:  [27.884, 11.425, 21.545]
alpha_fixed_emb: 0.95
asso_func: giou
aw_off: false
aw_param: 0.5
cmc_off: false
conf: 0.5
delta_t: 3
det_thresh: 0
embedding_off: false
inertia: 0.2
iou_thresh: 0.3
max_age: 30
min_hits: 1
new_kf_off: false
w_association_emb: 0.75

```

**Comment by mikel-brostrom** [comment]
If you check out the results in README they shouldn't. Is this what you see for your specific dataset?

**Comment by Jumabek** [comment]
@mikel-brostrom  No, This is on MOT17

Results in Readme are only for first 10 frames of mot17, is it not?

**Comment by mikel-brostrom** [comment]
I introduced a bug both in OCSORT and DeepOCSORT when adding support for segmentation and pose estimation models. This has now been fixed (https://github.com/mikel-brostrom/yolo_tracking/pull/1112)

**Comment by Jumabek** [comment]
Thanks @mikel-brostrom 

let me attempt reproducing results

**Comment by Jumabek** [comment]
I confirm, here is the command and output of HOTA metrics for MOT17 on YOLOX_m as a detector

```
# List of tracking methods
tracking_methods=("ocsort" "deepocsort" "bytetrack" "strongsort" "botsort")

# # Loop through each tracking method and run the command
for tracking_method in "${tracking_methods[@]}"; do
    python examples/val.py --yolo-model yolo_nas_m --tracking-method "$tracking_method" --benchmark MOT17 --classes 0
done

```


**Results**
OCSORT	                59.591
DeepOCSORT	60.612
StrongSORT	        61.124
ByteTrack	        60.998
BotSORT	                62.576

**Comment by mikel-brostrom** [comment]
Did you use the code in master for this @Jumabek ?

**Comment by Jumabek** [comment]
Yes,
Do you see any inconsistencies @mikel-brostrom

**Comment by mikel-brostrom** [comment]
This value:

https://github.com/mikel-brostrom/yolo_tracking/blob/774e13322f31931e42ace0f39e572f6c77046a00/boxmot/configs/deepocsort.yaml#L15

should be equal to 3 to mimic the original implementation:

https://github.com/GerardMaggiolino/Deep-OC-SORT/blob/6bb51d027b137233f5c520b6fcc4f2ae387a6ba9/trackers/ocsort_tracker/ocsort.py#L202

 It is set to 1 in this implementation for the sake of maximum performance on the first frames of each MOT17 sequence.

**Comment by mikel-brostrom** [comment]
But 

> I got higher performance for OCSORT against DeepOCSORT on MOT17

Is now solved. So  maybe we want to start a new issue? :smile:
